test/core/script.wast

```diff
@@ -0,0 +1,100 @@
+(module $m1
+  (global (export "g") i32 (i32.const 41))
+  (func (export "f") (result i32) (i32.const 41))
+  (func (export "inc") (param $x i32) (result i32)
+    (i32.add (local.get $x) (i32.const 1))
+  )
+)
+
+(module $m2
+  (global (export "g") (mut i32) (i32.const 42))
+  (func (export "f") (result i32) (i32.const 42))
+  (func (export "add3") (param i32 i32 i32) (result i32)
+    (i32.add (i32.add (local.get 0) (local.get 1)) (local.get 2))
+  )
+  (func (export "swap") (param i32 i32) (result i32 i32)
+    (local.get 1) (local.get 0)
+  )
+)
+
+(assert_return (get "g") (i32.const 42))
+(assert_return (get $m1 "g") (i32.const 41))
+(assert_return (get $m2 "g") (i32.const 42))
+
+(set "g" (i32.const 43))
+(assert_return (set "g" (i32.const 43)))
+(assert_return (get "g") (i32.const 43))
+(set $m2 "g" (i32.const 44))
+(assert_return (get "g") (i32.const 44))
+(set "g" (invoke $m1 "inc" (get "g")))
+(assert_return (get "g") (i32.const 45))
+(assert_return (get $m1 "g") (i32.const 41))
+(assert_return (get $m2 "g") (i32.const 45))
+
+(assert_return (invoke "f") (i32.const 42))
+(assert_return (invoke $m1 "f") (i32.const 41))
+(assert_return (invoke $m2 "f") (i32.const 42))
+
+(assert_return (invoke $m1 "inc" (i32.const 2)) (i32.const 3))
+(assert_return (invoke $m1 "inc" (get $m1 "g")) (i32.const 42))
+(assert_return (invoke $m1 "inc" (get $m2 "g")) (i32.const 46))
+(assert_return (invoke $m1 "inc" (invoke $m1 "inc" (get "g"))) (i32.const 47))
```

Comment:
I'm wondering if intentionally limiting the grammar to not allow nested invokes (or even `set`) would be good. One the one hand, it simplifies the test runner so it doesn't have to keep an internal stack, and also forces tests to flatten out their nested calls, saving their intermediate results in globals.

Reply:
Actually, I have often missed the ability to simply plug two function calls together for an assertion. Not being able to do that on script level requires contortions exactly like those from not being able to access a global. So it would be a pity to exclude it. It's just a trivial recursive expression evaluator after all.

The real complexity with this extension is in translating tests to JS. But that all has to do with the fact that synthesised Wasm modules for assertions now require an arbitrary number of imports, which isn't changed by limiting expression depth.

(Also, I ran into a couple of V8/node bugs, such as #1597, and worse, deterministic bus errors on my new Arm MacBook. :( )